File: src/main/java/com/saomc/saoui/events/RenderHandler.java
package com.saomc.saoui.events;

import com.saomc.saoui.SoundCore;
import com.saomc.saoui.config.OptionCore;
import com.saomc.saoui.neo.screens.IngameMenuGUI;
import com.saomc.saoui.renders.StaticRenderer;
import com.saomc.saoui.screens.death.DeathScreen;
import com.saomc.saoui.screens.ingame.IngameGUI;
import com.saomc.saoui.screens.menu.StartupGUI;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.gui.GuiGameOver;
import net.minecraft.client.gui.GuiIngameMenu;
import net.minecraft.client.gui.GuiMainMenu;
import net.minecraft.client.gui.GuiOptions;
import net.minecraft.client.gui.inventory.GuiContainerCreative;
import net.minecraft.client.gui.inventory.GuiInventory;
import net.minecraft.client.renderer.entity.RenderManager;
import net.minecraft.entity.EntityLivingBase;
import net.minecraftforge.client.event.GuiOpenEvent;
import net.minecraftforge.client.event.RenderLivingEvent;
import net.minecraftforge.client.event.RenderPlayerEvent;

import java.util.ArrayList;
import java.util.List;

@SideOnly(Side.CLIENT)
class RenderHandler {

    private static final List<EntityLivingBase> deadHandlers = new ArrayList<>();
    private static boolean menuGUI = true;

    static void checkingameGUI() {
        if (EventCore.mc.ingameGUI != null && !(EventCore.mc.ingameGUI instanceof IngameGUI))
            EventCore.mc.ingameGUI = new IngameGUI(EventCore.mc);
    }

    static void deathHandlers() {
        deadHandlers.forEach(ent -> {
            if (ent != null) {
                final boolean deadStart = (ent.deathTime == 1);
                final boolean deadExactly = (ent.deathTime >= 18);
                if (deadStart) {
                    ent.deathTime++;
                    SoundCore.playAtEntity(ent, SoundCore.PARTICLES_DEATH);
                }

                if (deadExactly) {
                    StaticRenderer.doSpawnDeathParticles(EventCore.mc, ent);
                    ent.setDead();
                }
            }
        });
        deadHandlers.removeIf(ent -> ent.isDead);
    }

    static void guiInstance(GuiOpenEvent e) {
        if (!OptionCore.BUGGY_MENU.isEnabled()) return;
        if (OptionCore.DEBUG_MODE.isEnabled()) System.out.print(e.gui + " called GuiOpenEvent \n");

        if (e.gui instanceof GuiIngameMenu) {
            if (!(EventCore.mc.currentScreen instanceof IngameMenuGUI)) {
                e.gui = new IngameMenuGUI();
            }
        }
        if (e.gui instanceof GuiInventory && !OptionCore.DEFAULT_INVENTORY.isEnabled()) {
            if (EventCore.mc.playerController.isInCreativeMode())
                e.gui = new GuiContainerCreative(EventCore.mc.thePlayer);
            else if (!(EventCore.mc.currentScreen instanceof IngameMenuGUI))
                e.gui = new IngameMenuGUI(/*(GuiInventory) EventCore.mc.currentScreen*/);
            else e.setCanceled(true);
        }
        if (e.gui instanceof GuiGameOver && (!OptionCore.DEFAULT_DEATH_SCREEN.isEnabled())) {
            if (!(e.gui instanceof DeathScreen)) {
                e.gui = new DeathScreen();
            }
        }
        if (e.gui instanceof IngameMenuGUI)
            if (EventCore.mc.currentScreen instanceof GuiOptions) {
                e.setCanceled(true);
                EventCore.mc.currentScreen.onGuiClosed();
                EventCore.mc.setIngameFocus();
            }

    }

    static void deathCheck() {
        if (EventCore.mc.currentScreen instanceof DeathScreen && EventCore.mc.thePlayer.getHealth() > 0.0F) {
            EventCore.mc.currentScreen.onGuiClosed();
            EventCore.mc.setIngameFocus();
        }
    }

    static void renderPlayer(RenderPlayerEvent.Post e) {
        if (!OptionCore.UI_ONLY.isEnabled()) {
            if (e.entityPlayer != null) {
                StaticRenderer.render(RenderManager.instance, e.entityPlayer, e.entityPlayer.posX, e.entityPlayer.posY, e.entityPlayer.posZ);
            }
        }
    }

    static void renderEntity(RenderLivingEvent.Post e) {
        if (!OptionCore.UI_ONLY.isEnabled()) {
            if (e.entity != EventCore.mc.thePlayer) {
                StaticRenderer.render(RenderManager.instance, e.entity, e.x, e.y, e.z);
            }
        }
    }

    static void mainMenuGUI(GuiOpenEvent e) {
        if (menuGUI)
            if (e.gui instanceof GuiMainMenu)
                if (StartupGUI.shouldShow()) {
                    e.gui = new StartupGUI();
                    menuGUI = false;
                } //else e.gui = new MainMenuGUI());
    }

}

File: src/main/java/com/saomc/saoui/events/RenderHandler.kt
/*
 * Copyright (C) 2016-2019 Arnaud 'Bluexin' Solé
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.saomc.saoui.events

import com.saomc.saoui.SoundCore
import com.saomc.saoui.capabilities.getRenderData
import com.saomc.saoui.config.OptionCore
import com.saomc.saoui.playAtEntity
import com.saomc.saoui.renders.StaticRenderer
import com.saomc.saoui.screens.CoreGUI
import com.saomc.saoui.screens.DeathGui
import com.saomc.saoui.screens.menus.IngameMenu
import net.minecraft.client.gui.GuiGameOver
import net.minecraft.client.gui.GuiIngameMenu
import net.minecraft.client.gui.GuiMainMenu
import net.minecraft.client.gui.inventory.GuiContainerCreative
import net.minecraft.client.gui.inventory.GuiInventory
import net.minecraft.entity.EntityLivingBase
import net.minecraftforge.client.event.GuiOpenEvent
import net.minecraftforge.client.event.RenderLivingEvent
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.util.*

@SideOnly(Side.CLIENT)
internal object RenderHandler {

    private val deadHandlers = ArrayList<EntityLivingBase>()
    private var menuGUI = true

    fun deathHandlers() {
        deadHandlers.forEach { ent ->
            val deadStart = ent.deathTime == 1
            val deadExactly = ent.deathTime >= 18
            if (deadStart) {
                ent.deathTime++
                SoundCore.PARTICLES_DEATH.playAtEntity(ent)
            }

            if (deadExactly) {
                StaticRenderer.doSpawnDeathParticles(EventCore.mc, ent)
            }
        }
        deadHandlers.removeIf { ent -> ent.deathTime >= 18 }
    }

    fun addDeadMob(ent: EntityLivingBase) {
        deadHandlers.add(ent)
    }

    fun guiInstance(e: GuiOpenEvent) {
        if (e.gui is GuiIngameMenu) {
                if (EventCore.mc.currentScreen !is CoreGUI<*>) e.gui = IngameMenu()
        }
        else if (e.gui is GuiInventory && !OptionCore.DEFAULT_INVENTORY.isEnabled) {
            when {
                EventCore.mc.playerController.isInCreativeMode -> e.gui = GuiContainerCreative(EventCore.mc.player)
                else -> e.isCanceled = true
            }
        }
        else if (e.gui is GuiGameOver && !OptionCore.DEFAULT_DEATH_SCREEN.isEnabled) {
            if (EventCore.mc.currentScreen !is DeathGui) e.gui = DeathGui()
            else {
                e.isCanceled = true
            }
        }
    }


    fun renderEntity(e: RenderLivingEvent.Post<*>) {
        if (!OptionCore.UI_ONLY.isEnabled) {
            e.entity.getRenderData()?.update(e.partialRenderTick)
        }
    }

    fun mainMenuGUI(e: GuiOpenEvent) {
        if (menuGUI)
            if (e.gui is GuiMainMenu) {}
            //e.setGui(new MainMenuGUI());
    }

}
